cpp/src/groupby/sort/sort_helper.cu

```diff
@@ -0,0 +1,312 @@
+
+    // All rows with one or more null values are at the end of the resulting sorted order.
+  }
+
+  return _key_sorted_order->view();
+}
+
+helper::index_vector const& helper::group_offsets(cudaStream_t stream) {
+  if (_group_offsets)
+    return *_group_offsets;
+
+  _group_offsets = std::make_unique<index_vector>(num_keys(stream));
+
+  auto device_input_table = table_device_view::create(_keys, stream);
+  auto sorted_order = key_sort_order().data<size_type>();
+  decltype(_group_offsets->begin()) result_end;
+  auto exec = rmm::exec_policy(stream);
+
+  if (has_nulls(_keys)) {
+    result_end = thrust::unique_copy(exec->on(stream),
+      thrust::make_counting_iterator<size_type>(0),
+      thrust::make_counting_iterator<size_type>(num_keys(stream)),
+      _group_offsets->begin(),
+      permuted_row_equality_comparator<true>(*device_input_table, sorted_order));
+  } else {
+    result_end = thrust::unique_copy(exec->on(stream), 
+      thrust::make_counting_iterator<size_type>(0),
+      thrust::make_counting_iterator<size_type>(num_keys(stream)),
+      _group_offsets->begin(),
+      permuted_row_equality_comparator<false>(*device_input_table, sorted_order));
+  }
+
+  size_type num_groups = thrust::distance(_group_offsets->begin(), result_end);
+  _group_offsets->resize(num_groups);
+
+  return *_group_offsets;
+}
+
+helper::index_vector const& helper::group_labels(cudaStream_t stream) {
+  if (_group_labels)
+    return *_group_labels;
+
+  // Get group labels for future use in segmented sorting
+  _group_labels = std::make_unique<index_vector>(num_keys(stream));
+
+  auto& group_labels = *_group_labels;
+
+  if (num_keys(stream) == 0)
+    return group_labels;
+
+  auto exec = rmm::exec_policy(stream);
+  thrust::scatter(exec->on(stream),
+    thrust::make_constant_iterator(1, decltype(num_groups())(1)), 
+    thrust::make_constant_iterator(1, num_groups()), 
+    group_offsets().begin() + 1, 
+    group_labels.begin());
+ 
+  thrust::inclusive_scan(exec->on(stream),
+                        group_labels.begin(),
+                        group_labels.end(),
+                        group_labels.begin());
+
+  return group_labels;
+}
+
+column_view helper::unsorted_keys_labels(cudaStream_t stream) {
+  if (_unsorted_keys_labels)
+    return _unsorted_keys_labels->view();
+
+  column_ptr temp_labels = make_numeric_column(
+                              data_type(type_to_id<size_type>()),
+                              _keys.num_rows(), 
+                              mask_state::ALL_NULL, stream);
+  
+  auto group_labels_view = cudf::column_view(
+                              data_type(type_to_id<size_type>()),
+                              group_labels().size(),
+                              group_labels().data().get());
+  
+  auto scatter_map = cudf::experimental::detail::slice(
+    key_sort_order(), 0, num_keys(stream));
+
+  std::unique_ptr<table> t_unsorted_keys_labels = 
+    cudf::experimental::detail::scatter(
```

Comment:
Which variant? I do see one for gather but not for scatter.

Reply:
Hm, doesn't look like it was ever created... I'll file an issue.

Reply:
https://github.com/rapidsai/cudf/issues/3745

Reply:
@jrhemstad There is no gather variant which uses an iterator instead of a column/table view. Did you mean only for scatter map? I talking with respective to the first comment that you gave here. I don't have any issues with requirements of #3745

Reply:
@rgsl888prabhu I believe he means this https://github.com/rapidsai/cudf/blob/1f81d7296e7db4fb7457ac7a945ef92371d7923d/cpp/include/cudf/detail/gather.cuh#L300-L307

Reply:
Ah, I got confused when `group_labels` which was being used as a column changed to be gather map.

Reply:
@jrhemstad @devavret https://github.com/rapidsai/cudf/pull/3758/files
Scatter variant is ready for review

Reply:
Since there is no copy involved in the current method, I'll consider this not a blocker. Unless #3758 is merged by Jan 14.